File: packages/gavio-js/src/interceptors/cache/embedding.ts
/**
 * Embeddings for the semantic cache (F-CACHE-02).
 *
 * Zero-dependency hashed bag-of-words embedder (L2-normalised) — good enough to
 * dedup near-identical prompts. Plug in a real embedder implementing `Embedder`
 * for production semantic matching.
 */

import { createHash } from 'node:crypto'

export interface Embedder {
  embed(text: string): number[]
}

const TOKEN = /[a-z0-9]+/g

/** Deterministic hashed bag-of-words embedder. */
export function hashingEmbedder(dim = 256): Embedder {
  return {
    embed(text: string): number[] {
      const vec = new Array<number>(dim).fill(0)
      const tokens = text.toLowerCase().match(TOKEN) ?? []
      for (const token of tokens) {
        // Parity note: Python uses blake2b(digest_size=8); here we take the
        // first 8 bytes of blake2b512. Both are deterministic; the JS cache is
        // per-process so cross-language byte-parity is not required.
        const digest = createHash('blake2b512').update(token).digest()
        let n = 0n
        for (let i = 0; i < 8; i++) n = (n << 8n) | BigInt(digest[i]!)
        const bucket = Number(n % BigInt(dim))
        vec[bucket]! += 1
      }
      const norm = Math.sqrt(vec.reduce((s, x) => s + x * x, 0))
      if (norm === 0) return vec
      return vec.map((x) => x / norm)
    },
  }
}

/** Cosine similarity; safe for zero vectors. */
export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length) throw new Error('vectors must have equal length')
  let dot = 0
  let na = 0
  let nb = 0
  for (let i = 0; i < a.length; i++) {
    dot += a[i]! * b[i]!
    na += a[i]! * a[i]!
    nb += b[i]! * b[i]!
  }
  if (na === 0 || nb === 0) return 0
  return dot / (Math.sqrt(na) * Math.sqrt(nb))
}

File: packages/gavio-py/gavio/interceptors/cache/embedding.py
"""Embeddings for the semantic cache (F-CACHE-02).

The core stays dependency-free, so it ships a deterministic hashing embedder
(hashed bag-of-words, L2-normalised) — good enough to dedup near-identical
prompts (case/whitespace/punctuation variants) with zero dependencies. For
production semantic matching, plug in a real embedder (e.g. an OpenAI
``text-embedding-3-small`` adapter) that implements the same ``Embedder``
protocol.
"""

from __future__ import annotations

import hashlib
import math
import re
from typing import Protocol, runtime_checkable

_TOKEN = re.compile(r"[a-z0-9]+")


@runtime_checkable
class Embedder(Protocol):
    """Turns text into a fixed-length float vector."""

    def embed(self, text: str) -> list[float]: ...


class HashingEmbedder:
    """Zero-dependency hashed bag-of-words embedder (L2-normalised)."""

    def __init__(self, dim: int = 256) -> None:
        self.dim = dim

    def embed(self, text: str) -> list[float]:
        vec = [0.0] * self.dim
        for token in _TOKEN.findall(text.lower()):
            digest = hashlib.blake2b(token.encode("utf-8"), digest_size=8).digest()
            bucket = int.from_bytes(digest, "big") % self.dim
            vec[bucket] += 1.0
        norm = math.sqrt(sum(x * x for x in vec))
        if norm == 0.0:
            return vec
        return [x / norm for x in vec]


def cosine_similarity(a: list[float], b: list[float]) -> float:
    """Cosine similarity. Assumes equal length; safe for zero vectors."""
    if len(a) != len(b):
        raise ValueError("vectors must have equal length")
    dot = sum(x * y for x, y in zip(a, b, strict=True))
    na = math.sqrt(sum(x * x for x in a))
    nb = math.sqrt(sum(y * y for y in b))
    if na == 0.0 or nb == 0.0:
        return 0.0
    return dot / (na * nb)
